File: Week1/2p_games/Nim/nim.py
n = int(input("How many rows do you want: "))

L = [2*i+1 for i in range(n)]

player1 = True
def display():
    for i in range(n):
        print((i+1),"   ","  "*(n-i-1),"* "*L[i],sep='')
def play():
    display()
    global L
    global player1
    if player1 == True:
        s = input("Enter Player1 move ")
    else:
        s = input("Enter Player2 move ")
    
    row_number,sticks_removed = list(map(int,s.strip().split("-")))

    while  row_number>0 and row_number<=n and L[row_number-1] < sticks_removed:
        print("Enter a valid move\n ")
        if player1 == True:
            s = input("Enter Player1 move ")
        else:
            s = input("Enter Player2 move ")
        row_number,sticks_removed = list(map(int,s.strip().split("-")))

    L[row_number-1]-=sticks_removed

    if L==[0]*n:
        if player1:
            print("Player1 wins")
        else:
            print("Player2 wins")
        exit(0)

    player1 = not player1
    play()

if __name__== '__main__':
    play()

File: Week1/2p_games/Nim/nim.cpp
#include<iostream>
#include<cstdlib>
using namespace std;
void computer_move(int * nums , int n)
{
    int Xor = 0;
    for (int i = 0 ; i < n ; i++) Xor ^= nums[i];
    
    if (Xor == 0)
    {
        // Computer can't win so just play a random move.
        for (int i = 0 ; i < n ; i++)
        {
            if (nums[i])
            {
                nums[i] = rand()%nums[i];
                break;
            }
        }
        return;
    }
    int k = 1;
    int XXor = Xor;
    while (Xor)
    {
        Xor >>= 1;
        k <<= 1;
    }
    k>>=1;
    
    // k finally becomes largest power of 2 in binary representation of variable Xor
    for (int i = 0; i < n ; i++)
    {
        if (nums[i] & k)
        {
            nums[i] = nums[i] ^ XXor;
            return;
        }
    }
}
bool end(int * nums , int n)
{
    for (int i = 0 ; i < n ; i++) 
    {
        if (nums[i]) return false;
    }
    return true;
}
int main()
{
    int n ;
    cout << "Enter the number of elements: ";
    cin >> n;
    int nums[n];
    cout << "Enter the elements";

    int Xor = 0;
    for (int i = 0 ; i < n ; i++)
    {
        cin >> nums[i];
        Xor ^= nums[i];
    }
    // if (Xor == 0)
    // {
    //     cout << "Player moving first loses";
    // }
    // else
    // {
    //     cout << "Player moving first wins";
    // }

    cout << "Do you want to play the first move (enter y/n): ";
    char input; bool player_move;
    cin >> input;

    if (input == 'y') player_move = true;
    else player_move = false;

    while (true)
    {
        if (player_move)
        {
            for (int i = 0 ; i < n ; i++)
            {
                cout << nums[i] << " ";  
            }
            cout << "\n";
            while (true)
            {
                cout << "Enter your move: ";
                int row , value;
                cin >> row >> value;
                if (row >0 && row <=n && value>=0 && nums[row-1] > value)
                {
                    nums[row-1] = value;
                    break;
                }
                else
                {
                    cout << "Invalid move\n";
                }
            }
            for (int i = 0 ; i < n ; i++)
            {
                cout << nums[i] << " ";  
            }
            cout << "\n";

            if (end(nums,n)) 
            {
                cout << "Congratulations!! You won";
                return 0;
            }
        }
        else
        {
            computer_move(nums,n);
            for (int i = 0 ; i < n ; i++)
            {
                cout << nums[i] << " ";  
            }
            cout << "\n";
            if (end(nums,n)) 
            {
                cout << "Sorry! You lost";
                return 0;
            }
        }
        player_move = !player_move;
    }

}